server/src/main/java/io/crate/planner/Planner.java

```diff
@@ -213,385 +218,414 @@ public ClusterState currentClusterState() {
-                "CrateDB has no transactions, so any `SET LOCAL` change would be dropped in the next statement.");
+        @Override
+        public Plan visitSetTransaction(AnalyzedSetTransaction setTransaction, PlannerContext context) {
+            LOGGER.info("'SET TRANSACTION' statement is ignored. CrateDB doesn't support transactions");
             return NoopPlan.INSTANCE;
-        } else {
-            return new SetSessionAuthorizationPlan(analysis, userManager);
         }
-    }
 
-    @Override
-    public Plan visitSetTransaction(AnalyzedSetTransaction setTransaction, PlannerContext context) {
-        LOGGER.info("'SET TRANSACTION' statement is ignored. CrateDB doesn't support transactions");
-        return NoopPlan.INSTANCE;
-    }
-
-    @Override
-    public Plan visitKillAnalyzedStatement(AnalyzedKill analysis, PlannerContext context) {
-        return new KillPlan(analysis.jobId());
-    }
+        @Override
+        public Plan visitKillAnalyzedStatement(AnalyzedKill analysis, PlannerContext context) {
+            return new KillPlan(analysis.jobId());
+        }
 
-    @Override
-    public Plan visitDeallocateAnalyzedStatement(AnalyzedDeallocate analysis, PlannerContext context) {
-        return NoopPlan.INSTANCE;
-    }
+        @Override
+        public Plan visitDeallocateAnalyzedStatement(AnalyzedDeallocate analysis, PlannerContext context) {
+            return NoopPlan.INSTANCE;
+        }
 
-    @Override
-    public Plan visitDiscard(AnalyzedDiscard discard, PlannerContext context) {
-        return NoopPlan.INSTANCE;
-    }
+        @Override
+        public Plan visitDiscard(AnalyzedDiscard discard, PlannerContext context) {
+            return NoopPlan.INSTANCE;
+        }
 
-    @Override
-    public Plan visitExplainStatement(ExplainAnalyzedStatement explainAnalyzedStatement, PlannerContext context) {
-        ProfilingContext ctx = explainAnalyzedStatement.context();
-        if (ctx == null) {
-            return new ExplainPlan(explainAnalyzedStatement.statement().accept(this, context), explainAnalyzedStatement.showCosts(), null);
-        } else {
-            Timer timer = ctx.createAndStartTimer(ExplainPlan.Phase.Plan.name());
-            Plan subPlan = explainAnalyzedStatement.statement().accept(this, context);
-            ctx.stopTimerAndStoreDuration(timer);
-            return new ExplainPlan(subPlan, explainAnalyzedStatement.showCosts(), ctx);
+        @Override
+        public Plan visitExplainStatement(ExplainAnalyzedStatement explainAnalyzedStatement, PlannerContext context) {
+            AnalyzedStatementVisitor<PlannerContext, Plan> planner = this;
+            VerboseOptimizerProgressTracker tracker = new VerboseOptimizerProgressTracker(
+                defaultTracker,
+                explainAnalyzedStatement.showCosts()
+            );
+            if (explainAnalyzedStatement.verbose()) {
+                planner = new TrackablePlanner(tracker);
+            }
```

Comment:
What do you think of instead keeping `Planner` a visitor instance, but instead either extend `PlannerContext` to have a tracer property, or introduce a `PlanVisitorContext` (or something like that) that wraps `PlannerContext` and the Tracer?

Replacing the full Visitor/Planner only to change the used tracker instance seems a bit much to me and the approach wouldn't scale for more different parameterizations

Reply:
I like your idea. It's definitely less invasive than the original approach. Done.